### DISCUSSION TITLE
Overhaul/Clean up/Refactor SConstruct/SConscript

### DISCUSSION THREAD

**Discussion by bryanwweber** [Cantera/enhancements#93] [Ideas]
The SCons build scripts are currently quite an octopus. Although they aren't very fragile, in the sense that they tend to work on every platform we need them to, it's a little hard to know where to change/add options as needed. I think there's room to do a few things:

* Update the style of the various files to be consistent with modern Python style (rename functions from camelCase to snake_case, etc.)
* Use modern Python 3 features such as `pathlib.Path`, type hinting, etc. See also #18 
* Reorganize the main SConstruct file to clearly delineate where features are checked/added etc.

The overall goal is to accomplish this simplification without changing or dropping support for any of the current platforms/compilers we support.

I think doing this would have a few benefits. First, it will make it easier to onboard new folks who need to modify the build script. Second, it will make it easier to add new features, such as [interface-specific installers](https://github.com/Cantera/enhancements/issues/75) (which will, in turn, make it easier to do things like have a Wheel package to install with pip). Third, it will simplify the code by using modern Python features. And fourth, it will allow developers to use modern Python tooling, such as static type checkers (which are built in to editors like VS Code).

This is a longer term project, I think. Since it's not high-priority I don't imagine being able to devote lots of dedicated time (depending on how much procrastinating I'm doing with grading). I'm opening this as an Idea to see if other people have feedback about whether this is worth it, and if so, other places where simplification can happen!

**Comment by speth** [comment]
> And fourth, it will allow developers to use modern Python tooling, such as static type checkers (which are built in to editors like VS Code).

I think what these tools will all struggle with is some of the SCons magic. This includes both implicit import of various SCons classes like `Environment` and `Configure`, and especially the behavior of commands like `Import('env', 'build', 'install')` in the `SConscript` files. I guess it probably wouldn't be that hard to make Pylance or some other specific linters aware of the SCons types, but I'm really not sure how you would deal with the `Import` function, since there's no indication of what parent `SConstruct` file would be calling the current `SConscript`.

You can always add a vote for https://github.com/SCons/scons/issues/3210.

**Reply by bryanwweber** [reply]
> I think what these tools will all struggle with is some of the SCons magic.

Yeah, this is a bit unfortunate. I think we can probably get used to ignoring those specific errors, or perhaps `# noqa` them. Of course, if adding types everywhere makes things harder to read, that won't help either, so it will take a balance...

> You can always add a vote for SCons/scons#3210.

Done! :smile:

**Reply by speth** [reply]
I'm :-1: on littering code with `# noqa`, especially for this purpose. I'd rather figure out how to adjust the Pylance configuration to stop complaining about undefined variables in `SCons*`, and add that as a recommendation for developers to enable.

**Reply by bryanwweber** [reply]
In Pylance, you can set:
```
    "python.analysis.diagnosticSeverityOverrides": {
        "reportUndefinedVariable": "none"
    }
```
which disables reporting the undefined variables that come from SCons magic. The downside is that other undefined variables are no longer reported. Fortunately, it's usually pretty easy to find those by running one of the SCons commands (`scons build` etc.), it's not like this is a long-running Django site or something where you might hit an undefined variable that's going to take the site down.

**Comment by bryanwweber** [comment]
@speth @ischoegl @decaluwe 

Bike-shedding time: What's your opinion of single vs. double quotes in Python code (for my purposes here, in the various SCons related files)?

My personal opinion is a pretty strong preference for double quotes as the default, and my plan in this pull request is to eventually change all the single quotes to double quotes (except where single quotes have to be used because of a double quote inside the string). Although I'm not advocating to apply Black to our code base, I agree with their justification for preferring double quotes whenever possible:

> Why settle on double quotes? They anticipate apostrophes in English text. They match the docstring standard described in PEP 257. An empty string in double quotes ("") is impossible to confuse with a one double-quote regardless of fonts and syntax highlighting used. On top of this, double quotes for strings are consistent with C which Python interacts a lot with. [Source](https://black.readthedocs.io/en/stable/the_black_code_style.html#strings)

**Reply by ischoegl** [reply]
In my most recent coding, I've settled on double-quotes (mostly due to a preference for single-quotes in error messages etc.), so I'm basically in agreement.

Beyond, I'm not very orthodox about conventions (as long as code is readable and avoids clutter). Also, it looks like I learned a new term that I may use for some of my own musings  :grinning:

**Reply by bryanwweber** [reply]
Thanks @ischoegl! I should have also added to my original post that I'm most interested in having a convention that we can put into the contributing documentation and ask people to follow for consistency throughout the (Python) codebase.

**Reply by ischoegl** [reply]
Makes sense! I guess then you could add 'put a whitespace after a comma' ;) ... overall some of the white-space conventions (before/after `=`, except for keywords, and `+-*/`) could be pointed out. This is indeed a bike-shed :wink:

**Reply by bryanwweber** [reply]
Yes, and when @speth and I talked about this last time, he argued that single letter variables are easier to read _without_ the space, for example:

```python
for i,j in zip(list_1, list_2):
    pass
```

I personally prefer the space even in this case, although I try to avoid editing this type of thing unless I'm touching that line anyways.

Indeed, there are 2 problems with bike-sheds: 1) They're usually a waste of time to argue about, consistency is (IMO) more important 2) if you want to really fix them in a legacy codebase you get a HUGE diff, which totally screws up the `git blame`. Ah well.

**Reply by ischoegl** [reply]
At least in Python, I’d argue that PEP’s give decent recommendations without being overly stringent. Linters don’t like single-letter variables (I know why but at the same time they derive from math notation, so overall this is inconclusive). 

Regarding old code, I believe the strategy used for `double` vs `doublereal` applies here also ... even if consistency never converges. On that note, we could also argue about doxygen comment styles to put another addition on the shed!

**Reply by ischoegl** [reply]
Fwiw, I think it makes a lot of sense to settle some of these details for new code (where contributed code can get fixed before a merge), but I’m hesitant to change old code for the reason you mentioned (unnecessarily large diffs).

**Reply by speth** [reply]
There's more than just unnecessarily large diffs to worry about. The biggest one is that it makes merging/rebasing pull requests that were started before the formatting change extremely difficult. Applying style updates only to modified parts of the code avoids this, since you would have run into merge conflicts anyway.

**Reply by bryanwweber** [reply]
@speth so I'm proposing here to standardize the build system on double quotes, even for lines that wouldn't otherwise be modified by the other changes I'm making. Do you object strongly to doing that?

Separately, I'm proposing to standardize the entire (Python) code base (IIRC other languages are not so sanguine about the different quote types) on double quotes, for all modified lines going forward. I can put that one in a separate discussion, if you want.

**Reply by speth** [reply]
For the build system, I think it's OK, because (a) I think you're probably changing enough code nearby that there won't be that many possible other changes that wouldn't cause merge conflicts anyway and (b) I doubt there are many in-progress modifications to the build system that would be affected anyway.

As far as standardizing on double quotes in Python, I think it would be fine to write that in the style guide, but I am concerned about how to enforce that in pull requests. I find that even with the limited style rules that we do try to enforce, it's challenging for comments about these sorts of issues not to drown out more substantive review questions about implementation and correctness, and depending on how authors choose to address such style issues, can lead to a really bloated commit history (especially if one uses the option to directly accept suggestions made in a review). I think this is already a problem, so maybe we should have a separate discussion on the topic.

_I'll leave out of this the difficulty of breaking my longstanding habit of using single quotes in Python, which is definitely going to take some effort._

**Reply by ischoegl** [reply]
Alright. My muscle memory just reminded me of my own habits about single- vs. double-quotes. If it's a short string without white spaces, I usually used single-quotes. If it is a sentence that includes whitespaces, I used double quotes. I guess this will take more effort from my part also ...

**Reply by speth** [reply]
@bryanwweber, @ischoegl: After having had this rule in place for a good while, I'd like to reassess it. I think my prediction that fiddling around with quotation marks (and insisting on it in various places) would drown out more important review considerations has indeed come to pass, and I think we are wasting time in leaving review comments and "fixing" what's not broken. Further, I have seen multiple cases where diffs are harder to read because of attempts to adhere to the practice of changing the quotation style when any other change is being made to that line.

As more minor points, I also realized that this preference for double quotes is opposite to the Python interpreter itself, which uses single quotes:
```py
>>> d = {"foo": 3, "bar": "baz"}
>>> d
{'foo': 3, 'bar': 'baz'}
```

And even Black itself has added a [`--skip-string-normalization`](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#command-line-options) option to just leave quotes alone.

**Reply by bryanwweber** [reply]
Thanks for the thoughtful feedback. Without a way to automate this, it's just noise as you say. I'm fine with dropping the requirement.

**Reply by ischoegl** [reply]
I have no issue with dropping the rule. I don’t have strong feelings one way or the other. Only bad thing is that in hindsight, we are left with the worst of all outcomes: being stuck more or less in the middle of a transition with no plan to resolve inconsistent formatting. 

Didn’t mean to hit a sore point when pointing this out in a review today btw.

**Reply by bryanwweber** [reply]
> resolve inconsistent formatting.

I think that unless and until we find an auto-formatter that works on all our code (for example, I'm not aware of a formatter for Cython), we're just going to have to live with inconsistencies 🤷‍♂️

**Reply by ischoegl** [reply]
SCons is the most egregious example. Thankfully, it’s only a handful of people who have to live with an aesthetic disaster 😆. But it’s really not an issue worth losing sleep over.

**Reply by speth** [reply]
Thanks, both. I'm fine with leaving double quotes as a recommendation in our style guide, but just not pushing it as part of PR reviews.

**Comment by bryanwweber** [comment]
Another thing I want to check before I go too much farther: @speth @decaluwe @ischoegl 

I'm planning to require Python 3.6 as the minimum version to _build_ Cantera (as far as I know, the Python interface still supports 3.5, that'd be a separate discussion IMO). Therefore, I'm using features like `f`-strings in the SCons* files (and associated imported files). Actually, I think that's the only thing I've used so far that's only in 3.6 and up.

Ubuntu 18.04 (the oldest currently supported Ubuntu) includes Python 3.6 in the apt repositories. Python 3.5 ended official support in September 2020, although it is more-than-likely still present on older systems. For those systems without OS-level Python 3.6, I'd advocate that you should not be using a system Python that is no longer supported (I'd advocate not using the system Python to begin with, but that's somewhat separate).

Please respond if you have a problem with bumping the minimum Python for building Cantera to 3.6. :smiley:

**Reply by speth** [reply]
Besides the "oldest supported Ubuntu" and the official PSF support, I think the other system to keep in mind for older Python versions is RHEL. I can't tell what version is available as an official RedHat software collection for RHEL7, but there is at least a [community-maintained collection for Python 3.6](https://www.softwarecollections.org/en/scls/rhscl/rh-python36/) (but not any newer versions). So I'm 👍 for bumping the minimum Python requirement. I'm not sure that it's worth the level of confusion to have this be different from the version requirement for the Python module, but we don't really need to decide that now.

**Reply by bryanwweber** [reply]
I've used a few RHEL 7 systems at UConn, I'm fairly certain they had official support for Python 3.6. So that's 👍, great

**Comment by ischoegl** [comment]
**Python testing** ...  as this is a discussion about a `scons` overhaul (which likely will take more than one PR), I wanted to suggest running the python test suite through `pytest`, which imho generates superior error stacks (when configured to provide verbose output). A significant advantage is that it becomes trivial to run individual tests:
```
pytest path/to/testfile.py::testModule::testName
```
It's fully compatible with `unittest`, so the churn should be manageable. The main question would be configuration of scons, where I assume that it can be done.

**Reply by bryanwweber** [reply]
I've also talked to @speth about this, it's on the long-term roadmap in my head. I think pytest offers a few other advantages, such as their fixtures and test parameterization. We've also talked about adding Hypothesis for property-based testing, to supplement the mostly regression testing we have now.

FWIW, you can run individual tests pretty easily for the existing `unittest` suite:

```shell
PYTHONPATH=build/python python -m unittest cantera.test.test_module[.test_class[.test_name]]
```

**Reply by bryanwweber** [reply]
Another advantage of Pytest is that I bet we can get that to run/manage the Googletest based test suite as well, for example: https://pypi.org/project/pytest-cpp/

**Reply by ischoegl** [reply]
True about the `unittest` command, but for some reason I was never able to memorize that one, and mostly didn't feel like hunting for the posts where it was mentioned (anything that starts with caps `PYTHONPATH` is mentally blocked!). `pyenv` would presumably make that unnecessary, but that's another matter.

**Reply by bryanwweber** [reply]
I use

```shell
PYTHONPATH=build/python ipython
```

all the time for testing changes in the Python interface. Saves having to install the interface somewhere etc. etc. I don't think I've ever posted it very many places publicly because, as you imply, `PYTHONPATH` can get you into trouble, but this only works for one-off commands, as opposed to `export`ing in the `.zshrc` or something, so it can't cause too much trouble 😄

**Reply by speth** [reply]
If there's a way to use `pytest` if it's installed but fall back to `unittest`, I have no objection. Otherwise, I'm not sure it's really worth adding another build dependency.

I also use `PYTHONPATH=build/python` all the time with either `ipython` or `python -m unittest ...`. And for the gtest suite, you can use commands like `scons test-general gtest_flags=--gtest_filter=AnyMap.missingKey`, which is a bit awkward and has the unfortunate effect of saving the value of `gtest_flags` in `cantera.conf` even when you don't want it to. A simpler / easier to remember command for running a single test out of either the Python or gtest test suites would probably be beneficial.

**Comment by ischoegl** [comment]
**Packaging** ... another comment I have is about packaging. There currently are msi, macOS, conda and ubuntu packages (I may be missing a few). The first can be handled via `scons msi`, whereas the others cannot. I am not sure whether it is widely used, so it may be just be the 9-th arm of the octopus.

*Tangent:* Assuming that local installations are possible (I have *not* checked), a `scons conda` (to build a conda package) along with instructions would imho be more useful. It's probably still better to keep it outside of the main build script though.

 [*Edit: just noticed that [interface-specific-installers](https://github.com/Cantera/enhancements/issues/75) are already mentioned at the top*]

**Reply by bryanwweber** [reply]
> There currently are msi, macOS, conda and ubuntu packages (I may be missing a few). The first can be handled via `scons msi`, whereas the others cannot.

I wonder if it'd make sense to go the other direction and pull `scons msi` out into a separate repo, along with the macOS package (essentially combine the two packaging repos we already have for those interfaces). In addition, I'm hoping to add a Wheel-based installer (upon which, we could probably deprecate the Python MSI packages).

> _Tangent:_ Assuming that local installations are possible (I have _not_ checked), a `scons conda` (to build a conda package) along with instructions would imho be more useful

I think this would also be better going the other way. Doing `scons conda` would require shelling out to the `conda build` process (or figuring out their internal API), whereas adding a `scons install-python` would make the existing conda build scripts somewhat simpler.

**Reply by ischoegl** [reply]
> I wonder if it'd make sense to go the other direction and pull `scons msi` out into a separate repo

:+1: I guess this was my (implied) thought here.

> I think this would also be better going the other way. 

:+1: Anything that makes things simpler is good. Maybe there's a way to create a single packaging repo, which can dispatch via GitHub actions, but also provides instructions to build msi, conda, etc locally?

**Reply by bryanwweber** [reply]
> Maybe there's a way to create a single packaging repo, which can dispatch via GitHub actions, but also provides instructions to build msi, conda, etc locally?

I think that's a great idea, maybe one day 😄

**Reply by ischoegl** [reply]
A bike shed worthy of a king, I know ...

**Reply by speth** [reply]
At this point, I think the MSI installer is really only useful as a means of installing the Matlab toolbox, and I'm not sure that it's much better than just providing a `.zip` file that users can extract wherever they want. We also now have the `cantera-matlab` Conda package, which is easier to use in at least some ways (not requiring users to fiddle around with setting paths).

In principle, the MSI also provides library and header files if the user wants to compile apps and link to Cantera that way, but I'm not sure that's ever been super effective, since our MSI only has libraries for a single version of MSVC. And again, those might  be just as efficiently distributed as a .zip file. 

I think some of the things that we relied on having this installer for have become less relevant, with the move away from the executable `ck2cti` program several versions ago and less reliance on having a central location for Cantera's built-in data files with the YAML format (which no longer relies on an equivalent of the old `elements.xml`).

**Comment by bryanwweber** [comment]
The first pull request targeting this discussion is Cantera/cantera#1012 and is ready for review :tada:

**Comment by ischoegl** [comment]
As another comment that may fit in this context: how hard would it be to split `libcantera` into multiple libraries for *core* and *applications*? (similar on the Python front end.) Some users may only be interested in linking to the species, reactions, etc. portion (and YAML parsing of course!!), but may not require reactor networks, one-dimensional simulations etc..

**Reply by bryanwweber** [reply]
What's the benefit of doing this?

**Reply by ischoegl** [reply]
Cantera is quite large, and mixes *core* and *application* objects at the same level. At the moment, we use
```Python
import cantera as ct
...
```
and then pick from `ct`. As an alternative, you could envision
```Python
from cantera.core import Solution
from cantera.zerod import Reactor
...
```
which *may* be more intuitive/instructive to new users. (I know this is not the exact same argument as the one at the top, but would be a logical extension.) I'm not saying that it should be done, of course. But if this finds traction, it would fit in the SConstruct/SConscripts overhaul. And, yes, I do realize that you could simply write `from cantera import Solution, Reactor`, but my point is that those are really two different contexts :grinning:

**Reply by bryanwweber** [reply]
IMO this is part of a much larger discussion about restructuring the Python package (or even the entirety of Cantera). I don't think it'd be sufficient to just change SConscript to get this to happen... This is something that's come up other places, particularly if/as Cantera expands to support new communities (e.g., electrochemistry) where having an applications "toolbox" would be very useful.

**Reply by ischoegl** [reply]
Very true that this is a much larger discussion. One of my pet peeves are clear/intuitive interfaces, so I'm happy that this has come up at other places. 

I still believe that `SConscript` is the right place for implementing this, as having separate repos for toolboxes is imho inherently inferior. There should be 'official' toolboxes that should be provided as turn-key solutions?

**Reply by bryanwweber** [reply]
I didn't mean separate repos, necessarily, but certainly (for the Python interface) all of the `.pyx` files would need to be restructured to support the modified import `cantera.core` etc. That's why I think it's not so much a build-management issue via SConscript, but rather directly with the interface files. The necessary changes to support a modified interface like that would be handled by moving files around and modifying `setup.py.in` (with possibly minimal changes to SConscript to find the moved files).

**Reply by ischoegl** [reply]
It is certainly true that the Python API can be restructured outside of `SConscript` (which is, as you point out, much simpler), but I believe a _clean_ solution would mirror what starts at the C++ level. A `core` C++ cantera library would be much leaner than application toolboxes (no sundials, etc.), which may be attractive for those working with OpenFOAM, or any other 3rd party application. It would be a long-term aspirational project, but I wanted to bring it up.

**Reply by bryanwweber** [reply]
> It would be a long-term aspirational project, but I wanted to bring it up.

I think it'd be worth having a separate discussion topic devoted to this idea 😄

**Reply by speth** [reply]
If one of you wants to start that, I can provide some of the context for why things are the way they are -- not all of these things are as easy to split up as you might like.

**Reply by ischoegl** [reply]
Will do.

**Reply by ischoegl** [reply]
Done - see #97

**Comment by bryanwweber** [comment]
I read today about Ralf Gommers' efforts to transition SciPy to the [Meson build system](https://mesonbuild.com/). His motivations are varied, but the main one was that distutils is deprecated in Python 3.10 and will be removed in 3.12, some 2 years from now, and NumPy (and hence SciPy) depend heavily on distutils. Setuptools is planning to include distutils with a cleaner API, but they haven't done so yet.

I wonder if something like this makes sense for Cantera too. We currently don't use distutils to drive the compiling like NumPy does, so we don't have to worry about the impending removal of distutils. There's also always the danger of missing corner cases that are fixed in the SCons build system, especially for platforms we don't test that often. Meson does seem very fast though, and as the basis of this discussion is the cruft in our SCons system, I thought it'd be worth considering a change of build system.

More information: https://labs.quansight.org/blog/2021/07/moving-scipy-to-meson/ https://github.com/scipy/scipy/issues/13615

**Reply by speth** [reply]
I don't think speed is a very important motivation for changing Cantera's build system right now. A decent estimate of the time used by SCons is what it takes to run `scons build` when everything is up to date, which on my system takes around 2 seconds, compared to a couple of minutes spent actually compiling source files if you have to do a full rebuild (which no build system is going to reduce).

If Meson had been around when I initiated the transition from the old make-based build system (and a separately "maintained" Visual Studio project), I would certainly have considered it, but I don't see any killer features to motivate switching now. I'm a little surprised that SCons didn't even bear mentioning in the discussion of viable build systems for SciPy.

**Reply by bryanwweber** [reply]
I agree, those are all good points. SCons handles parallel builds where distutils doesn't, which I think was a big part of the speedup.

I also thought it was interesting that SCons didn't merit a mention, other than a related package in passing. I wonder if he was focusing on PEP517 compatible builders?

I think one other feature that Meson has that might be useful is it's intentionally not very extensible. _If_ it can do what we need without crazy hack-arounds, that would make the system easier to maintain. That's a big "if" though, and it would probably take more than a little work to figure out where the problems are.
